File: FE/canonicalize_span.py
"""Canonicalizer For Spans

"""


from FE import canonicalize
from FE import cwast
from FE import type_corpus
from FE import typify
from FE import eval

############################################################
# Convert Spans to equvalent struct
#
# span mut u8 -> struct {pointer ptr mut  u8, len uint}
############################################################
SLICE_FIELD_POINTER = cwast.NAME.Make("pointer")
SLICE_FIELD_LENGTH = cwast.NAME.Make("length")


def _MakeSpanReplacementStruct(span_type: cwast.CanonType,
                               tc: type_corpus.TypeCorpus) -> cwast.DefRec:
    fields = [
        (SLICE_FIELD_POINTER, tc.InsertPtrType(
            span_type.mut, span_type.underlying_type())),
        (SLICE_FIELD_LENGTH,  tc.get_uint_canon_type())
    ]
    return typify.MakeDefRec(f"xtuple_{span_type.name}", fields, tc, cwast.SRCLOC_GENERATED)


def MakeAndRegisterSpanTypeReplacements(tc: type_corpus.TypeCorpus) -> list[cwast.DefRec]:
    """For all types directly involving spans, produce a replacement type
    and return the map from one the other

    Note: recs need no replacement because the type name of a rec does no depend on the
          fields. So if a field contains a span we can just update that type which we do
          in ReplaceSpans()
    """
    tc.ClearReplacementInfo()
    # Go through the type table in topological order.
    # Note: we add new types to the map while iterating over it so we take
    # a snapshot first
    out = []
    for ct in tc.topo_order[:]:
        if ct.desugared:
            continue
        if ct.is_span():
            rec = _MakeSpanReplacementStruct(ct, tc)
            out.append(rec)
            new_ct = rec.x_type
        # Rec and DefType are handled are not replaced.
        # Their names do not reflect their children.
        # In the case of DefRec we patch up the RecFields separately.
        # DefType is handled further down
        elif ct.node in (cwast.DefRec, cwast.DefType):
            continue
        else:
            new_ct = tc.MaybeGetReplacementType(ct)
            if not new_ct:
                continue
        ct.LinkReplacementType(new_ct)
    # Note: the DefType can violate topplogical ordering
    # with respect to the underlying type in the presence of generic modules.
    # So we handle it in a sepatate loop
    # (For the same reason we use two phase for creating a wrapped type.)
    for ct in tc.topo_order[:]:
        if ct.node is cwast.DefType:
            ut = ct.underlying_type()
            if ut.node not in (cwast.DefRec, cwast.DefType):
                new_ct = tc.MaybeGetReplacementType(ut)
                if new_ct:
                    ct.children[0] = new_ct
    return out


def _MakeValRecForSpan(pointer, length, span_rec: cwast.CanonType, srcloc) -> cwast.ValCompound:

    pointer_field, length_field = span_rec.ast_node.fields
    return canonicalize.MakeValCompound(span_rec,
                                        [(pointer_field.x_type, pointer),
                                         (length_field.x_type, length)], srcloc)


def FunReplaceSpanCastWithSpanVal(node, tc: type_corpus.TypeCorpus):
    """Eliminate Array to Span casts. """
    uint_type: cwast.CanonType = tc.get_uint_canon_type()

    def replacer(node, _parent):
        nonlocal tc, uint_type
        if isinstance(node, cwast.ExprAs):
            if (node.x_type != node.expr.x_type and
                node.x_type.is_span() and
                    node.expr.x_type.is_vec()):
                return canonicalize.MakeValSpanFromArray(
                    node.expr, node.x_type, tc, uint_type)
        return None

    cwast.MaybeReplaceAstRecursivelyWithParentPost(node, replacer)


def ReplaceSpans(node):
    """
    Replaces all span<X> expressions with rec named xtuple_span<X>
    (cast to spans are eliminated by ReplaceExplicitSpanCast)
    This should elminate all of ExprSizeOf and ExprOffsetOf as a side-effect

    Complications:
     TODO: see unused _ConvertMutSpanValRecToSpanValRec helper
     `span<u8> = span-mut<u8>` is ok before the change to structs but not afterwards
    """
    def replacer(node):
        if isinstance(node, cwast.ExprFront):
            cont_ct = node.container.x_type
            if cont_ct.is_rec():
                # get the pointer field from the rec that now represents the span
                # because of the post-order traversal, node.container has already been processed
                assert cont_ct.original_type.is_span()
                pointer_field = cont_ct.get_rec_field(0)
                return canonicalize.MakeExprField(node.container,  pointer_field, node.x_srcloc)

        # len of array is constant and should have already been eliminated
        elif isinstance(node, cwast.ExprLen):
            cont_ct = node.container.x_type
            if cont_ct.is_rec():
                # get the length field from the rec that now represents the span
                # because of the post-order traversal, node.container has already been processed
                assert cont_ct.original_type.is_span()
                length_field = cont_ct.get_rec_field(1)
                return canonicalize.MakeExprField(node.container, length_field, node.x_srcloc)

        if cwast.NF.TYPE_ANNOTATED in node.FLAGS:
            rec_ct = node.x_type.replacement_type
            if rec_ct is not None:
                if isinstance(node, (cwast.DefVar, cwast.DefGlobal, cwast.DefFun,
                                     #
                                     cwast.Expr3, cwast.ExprAddrOf, cwast.ExprCall,
                                     cwast.ExprDeref, cwast.ExprField, cwast.ExprNarrow,
                                     cwast.ExprStmt, cwast.ExprUnionUntagged, cwast.ExprUnwrap,
                                     cwast.ExprWiden,
                                     #
                                     cwast.FunParam, cwast.Id, cwast.RecField, cwast.TypeAuto,
                                     #
                                     cwast.ValAuto, cwast.ValPoint, cwast.ValCompound)):
                    typify.NodeChangeType(node, rec_ct)
                    return None
                elif isinstance(node, cwast.ValSpan):
                    return _MakeValRecForSpan(node.pointer, node.expr_size, rec_ct, node.x_srcloc)

                cwast.CompilerError(
                    node.x_srcloc, "do not know how to convert span related node " +
                    f"[{rec_ct.name}]: {type(node)} of type {node.x_type}")
        return None

    cwast.MaybeReplaceAstRecursivelyPost(node, replacer)

File: FE/canonicalize_span.cc

#include "FE/canonicalize_span.h"

#include <array>

#include "FE/canonicalize.h"
#include "FE/cwast_gen.h"
#include "FE/eval.h"
#include "FE/typify.h"
#include "Util/parse.h"

namespace cwerg::fe {

Node MakeSpanReplacementStruct(CanonType span_ct, TypeCorpus* tc) {
  std::array<NameAndType, 2> fields = {{
      NameAndType{NameNew("pointer"), tc->InsertPtrFromSpan(span_ct)},
      NameAndType{NameNew("length"), tc->get_uint_canon_type()},
  }};
  std::string name = "xtuple_";
  name += NameData(CanonType_name(span_ct));
  return MakeDefRec(NameNew(name), fields, tc);
}

void MakeAndRegisterSpanTypeReplacements(TypeCorpus* tc, NodeChain* out) {
  tc->ClearReplacementInfo();
  CanonType new_ct;
  for (CanonType ct : tc->InTopoOrder()) {
    if (CanonType_is_span(ct)) {
      Node rec = MakeSpanReplacementStruct(ct, tc);
      out->Append(rec);
      new_ct = Node_x_type(rec);
    } else {
      new_ct = tc->MaybeGetReplacementType(ct);
      if (new_ct.isnull()) {
        continue;
      }
    }
    CanonTypeLinkReplacementType(ct, new_ct);
  }
  // TODO:
  for (CanonType ct : tc->InTopoOrder()) {
    if (CanonType_is_wrapped(ct)) {
      CanonType unwrapped = CanonType_underlying_type(ct);
      new_ct = tc->MaybeGetReplacementType(unwrapped);
      if (!new_ct.isnull()) {
        CanonType_children(ct)[0] = new_ct;
      }
    }
  }
}

Node MakeValUndef(const SrcLoc& sl) {
  Node out = NodeNew(NT::ValUndef);
  NodeInitValUndef(out, kStrInvalid, sl);
  Node_x_eval(out) = kConstUndef;
  return out;
}

Node MakeValPoint(Node field_val, CanonType ct, const SrcLoc& sl) {
  Node out = NodeNew(NT::ValPoint);
  NodeInitValPoint(out, field_val, MakeValUndef(sl), kStrInvalid, sl, ct);
  Node_x_eval(out) = Node_x_eval(field_val);
  return out;
}

Node MakeValRecForSpan(Node pointer, Node length, CanonType replacement_ct,
                       const SrcLoc& sl) {
  Node field_pointer = Node_fields(CanonType_ast_node(replacement_ct));
  Node field_length = Node_next(field_pointer);

  NodeChain fields;
  fields.Append(MakeValPoint(pointer, Node_x_type(field_pointer), sl));
  fields.Append(MakeValPoint(length, Node_x_type(field_length), sl));

  Node out = NodeNew(NT::ValCompound);
  NodeInitValCompound(out, MakeTypeAuto(replacement_ct, sl), fields.First(),
                      kStrInvalid, sl, replacement_ct);
  return out;
}

void ReplaceSpans(Node mod) {
  auto replacer = [](Node node, Node parent) -> Node {
    if (node.kind() == NT::ExprLen) {
      CanonType rec_ct = Node_x_type(Node_container(node));
      if (CanonType_kind(rec_ct) == NT::DefRec) {
        Node def_rec = CanonType_ast_node(rec_ct);
        Node len_field = Node_next(Node_fields(def_rec));
        ASSERT(len_field.kind() == NT::RecField, "");
        const SrcLoc& sl = Node_srcloc(node);
        Node container = Node_container(node);
        NodeFree(node);
        return MakeExprField(container, len_field, sl);
      }
    }

    if (node.kind() == NT::ExprFront) {
      CanonType rec_ct = Node_x_type(Node_container(node));
      if (CanonType_kind(rec_ct) == NT::DefRec) {
        Node def_rec = CanonType_ast_node(rec_ct);
        Node pointer_field = Node_fields(def_rec);
        ASSERT(pointer_field.kind() == NT::RecField, "");
        const SrcLoc& sl = Node_srcloc(node);
        Node container = Node_container(node);
        NodeFree(node);
        return MakeExprField(container, pointer_field, sl);
      }
    }

    if (!NodeHasField(node, NFD_X_FIELD::type)) {
      return node;
    }

    CanonType ct = Node_x_type(node);
    CanonType replacement_ct = CanonType_replacement_type(ct);
    if (replacement_ct.isnull()) {
      return node;
    }
    switch (node.kind()) {
      case NT::DefFun:
      case NT::DefGlobal:
      case NT::DefVar:
      //
      case NT::Expr3:
      case NT::ExprAddrOf:
      case NT::ExprCall:
      case NT::ExprDeref:
      case NT::ExprField:
      case NT::ExprNarrow:
      case NT::ExprStmt:
      case NT::ExprUnionUntagged:
      case NT::ExprUnwrap:
      case NT::ExprWiden:
      //
      case NT::FunParam:
      case NT::Id:
      case NT::RecField:
      case NT::TypeAuto:
      //
      case NT::ValAuto:
      case NT::ValPoint:
      case NT::ValCompound:
        NodeChangeType(node, replacement_ct);
        return node;
      case NT::ValSpan: {
        Node out = MakeValRecForSpan(Node_pointer(node), Node_expr_size(node),
                                     replacement_ct, Node_srcloc(node));
        NodeFree(node);
        return out;
      }
      default:
        ASSERT(false, "Unhandled span replacement for node kind "
                          << EnumToString(node.kind()));
        return node;
    }
  };

  MaybeReplaceAstRecursivelyPost(mod, replacer, kNodeInvalid);
}

}  // namespace cwerg::fe